MvtWatermark/MvtWatermark/QimMvtWatermark/QimMvtWatermark.cs

```diff
@@ -0,0 +1,583 @@
+        var t = new Tile(tile.TileId);
+        var envelopeTile = CoordinateConverter.TileBounds(t.X, t.Y, t.Zoom);
+        envelopeTile = CoordinateConverter.DegreesToMeters(envelopeTile);
+
+        var a = envelopeTile.Height / _options.M;
+        var extentDist = envelopeTile.Height / _options.Extent;
+
+        var winx = GenerateWinx(key);
+        var map = GenerateMap(key);
+
+        var bits = new BitArray(_options.Nb, false);
+
+        var dictGeneralExtractionMethod = new Dictionary<int, Tuple<int, int>>(_options.Nb);
+        var dict = new Dictionary<int, int>(_options.Nb);
+
+        for (var i = 0; i < _options.Nb; i++)
+        {
+            dictGeneralExtractionMethod.Add(i, new(0, 0));
+            dict.Add(i, 0);
+        }
+
+        for (var i = 0; i < _options.M; i++)
+            for (var j = 0; j < _options.M; j++)
+            {
+
+                var index = winx[i, j];
+                if (index == -1)
+                    continue;
+
+                var polygon = new Polygon(
+                    new LinearRing(
+                        new Coordinate[]
+                        {
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * j),
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * (j + 1)),
+                                    new(envelopeTile.MinX + a * (i + 1), envelopeTile.MinY + a * (j + 1)),
+                                    new(envelopeTile.MinX + a * (i + 1), envelopeTile.MinY + a * j),
+                                    new(envelopeTile.MinX + a * i, envelopeTile.MinY + a * j)
+                        }
+                ));
+
+                var stat = Statistics(tile, polygon, map, envelopeTile, extentDist, out var s0, out var s1);
+                if (Math.Abs(stat + 1) < 0.00001)
+                    continue;
+
+                dictGeneralExtractionMethod[index] = new(dictGeneralExtractionMethod[index].Item1 + s0, dictGeneralExtractionMethod[index].Item2 + s1);
+
+                embedded = true;
+                if (stat >= _options.T2)
+                {
+                    if (s0 > s1)
+                        dict[index] -= 1;
+
+                    if (s1 > s0)
+                        dict[index] += 1;
+                }
+            }
+
+        if (_options.IsGeneralExtractionMethod)
+        {
+            for (var i = 0; i < _options.Nb; i++)
+            {
+                var s0 = dictGeneralExtractionMethod[i].Item1;
+                var s1 = dictGeneralExtractionMethod[i].Item2;
+                if ((double)Math.Abs(s0 - s1) / (s1 + s0) > _options.T2)
+                {
+                    if (s1 > s0)
+                        bits[i] = true;
+                    if (s0 > s1)
+                        bits[i] = false;
+                }
+            }
+        }
+        else
+        {
+            for (var i = 0; i < _options.Nb; i++)
+            {
+                if (dict[i] > 0)
```

Comment:
Majority voting is not implemented (see the final item in the description of the extraction procedure)